lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,977 @@
+	}, nil
+}
+
+// Start starts the remote signer client. The function will continuously try to
+// setup a connection to the configured watch-only node, and retry to connect if
+// the connection fails until we Stop the remote signer client.
+func (r *OutboundClient) Start() error {
+	r.wg.Add(1)
+
+	// We'll continuously try setup a connection to the watch-only node, and
+	// retry to connect if the connection fails until we Stop the remote
+	// signer client.
+	go func() {
+		defer r.wg.Done()
+
+		for {
+			err := r.run()
+			if err != nil {
+				log.Errorf("Remote signer client error: %v",
+					err)
+			}
+
+			select {
+			case <-r.quit:
+				return
+			default:
+				log.Infof("Will retry to connect to "+
+					"watch-only node in: %v",
+					r.retryTimeout)
+
+				// Backoff before retrying to connect to the
+				// watch-only node.
+				select {
+				case <-r.quit:
+					return
+				case <-time.After(r.retryTimeout):
+				}
+			}
+
+			log.Infof("Retrying to connect to watch-only node")
+
+			// Increase the retry timeout by 50% for every retry.
+			r.retryTimeout = time.Duration(float64(r.retryTimeout) *
+				retryMultiplier)
+
+			// But cap the retryTimeout at r.maxRetryTimeout
+			if r.retryTimeout > r.maxRetryTimeout {
+				r.retryTimeout = r.maxRetryTimeout
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops the remote signer client.
+func (r *OutboundClient) Stop() error {
+	if r.stopped.Swap(true) {
+		return errors.New("remote signer client is already shut down")
+	}
+
+	log.Info("Remote signer client shutting down")
+
+	// Ensure that no new Run goroutines can start when we've initiated
+	// the stopping of the remote signer client.
+	r.wgMu.Lock()
+
+	if r.streamFeeder != nil {
+		r.streamFeeder.Stop()
+	}
+
+	close(r.quit)
+
+	r.wgMu.Unlock()
+
+	r.wg.Wait()
+
+	log.Debugf("Remote signer client shut down")
+
+	return nil
+}
+
+// run creates a new stream to the watch-only node, and starts processing and
+// responding to the sign requests that are sent over the stream. The function
+// will continuously run until the remote signer client is either stopped or
+// the stream errors.
+func (r *OutboundClient) run() error {
+	r.wgMu.Lock()
+
+	select {
+	case <-r.quit:
+		r.wgMu.Unlock()
+		return ErrShuttingDown
+	default:
+	}
+
+	r.wgMu.Unlock()
+
+	streamCtx, cancel := context.WithCancel(context.Background())
+
+	// Cancel the stream context whenever we return from this function.
+	defer cancel()
+
+	log.Infof("Attempting to setup the watch-only node connection")
+
+	// Try to get a new stream to the watch-only node.
+	stream, streamCleanUp, err := r.streamFeeder.GetStream(streamCtx)
+	if err != nil {
+		return err
+	}
+
+	r.stream = stream
+	defer streamCleanUp()
+
+	// Once the stream has been created, we'll need to perform the handshake
+	// process with the watch-only node, before it will start sending us
+	// requests.
+	err = r.handshake(streamCtx)
+	if err != nil {
+		return err
+	}
+
```

Comment:
it's strange to me that we go set `r.stream` just to go read it in `handshake` later. Seems like something that does not need to be a struct member.

Reply:
especially given the comment about always canceling the stream if we return from this. So that signals to me that it should explicitly be passed around and not be a long lived struct member.

Reply:
Changed so that `stream` isn't a struct member :)